### PR TITLE
Cleanup spec reload logic.

### PR DESCRIPTION
**Pull request by ferringb**
Due to the intertwined nature, this is one change that is essentially
two.

From the spec side, the mtime comparisons for key/cert against spec have
been removed- x509 comparisons are a far better approach, and reinforcing those
pathways is what we should invest in.

Unfortunately, cfssl still will not regenerate a key that has changed algo/keysize,
but that's a pre-existing issue- and one that will be fixed in follow up PRs.

These changes require shifting more internals of Spec to actual internal logic,
which leads to the manager changes.

The manager now asks the spec if there are changes on disk; if there are changes,
the manager in turn creates a new spec instance.  If that instance parses, we
use that- replacing the old.  If that spec doesn't parse, we use the old instance
already in memory.

This ensures that we don't corrupt the instance in memory from a failed loading of
a changed spec on disk.

Finally, the manager changes codify a subtle fact that certmgr doesn't properly
watch the directory of spec's, thus was never able to detect new content beyond
by chance.  This is fixable via integration of fsnotify, but that's something
for a later PR- in the interim, this watches just the spec's it saw at startup
and does reload of just those.

Per the comments, there is followup work; this commit doesn't make existing
behaviour any worse than it was already (for what's already bad), but does
fix behaviour where we can- and positions the code so those issues can be
more easily addressed.

### CODE REVIEW

File: cert/spec.go [hunk 1]
```diff
@@ -138,9 +141,15 @@ func newSpecFromPath(path string) (*Spec, error) {
 		return nil, err
 	}
 
+	specStat, err := os.Stat(path)
+	if err != nil {
+		// Hit the race; we read the file but someone wiped it.
+		return nil, err
```
Comment (willbuckner) [inline cert/spec.go]: I realize this probably applies to many other places in the repo, but its nice to add context to errors (very useful in debugging), for example: `errors.Wrap("newSpecFromPath: failed to stat cert spec", err)`

File: cert/spec.go [hunk 2]
```diff
@@ -262,34 +271,24 @@ func (spec *Spec) Lifespan() time.Duration {
...
-	st, err := os.Stat(path)
 	if err != nil {
 		if os.IsNotExist(err) {
-			log.Errorf("cert isChangedOnDisk: while checking path %s, got path error %s", path, err)
+			log.Debugf("spec %s was removed from on disk", spec)
+			return true, false, nil
 		}
-		return true
+		return false, false, err
+	} else if specStat.ModTime().After(spec.loadTime) {
```
Comment (willbuckner) [inline cert/spec.go]: nit: The `else` is unnecessary here, and conventional Go style for checking `err`  is to drop the else, add vertical space, and then put the next branch.

File: cert/spec.go (hunk 1 above)
Comment (ferringb) [reply cert/spec.go]: I'm viewing this as equivalent to python's `raise from` for exception chaining.  Correct me if I'm wrong please.

I'm a bit wary of attaching a stack trace to it, but the exception 'bubbling' up that occurs is already wrapping.

If the error dumping doesn't include tracebacks by default, I'm interested- if it does, I'm interested in ways to turn that off.

File: cert/spec.go (hunk 2 above)
Comment (ferringb) [reply cert/spec.go]: go linter don't care about that, so I'm undecided if I should care (if it was a raw 'if else', meh- I get their reasons but it's dogma in my view, but this isn't quite that).  Either way....

I'll be back in this pathway soon enough- the logic here assumes that mtime means "it's changed"; that doesn't, the content may be the same.  I was intending on adding parse comparisons to properly handle this in a follow up PR.  Will revisit it then.

File: cert/spec.go (hunk 1 above)
Comment (willbuckner) [reply cert/spec.go]: Go doesn't attach context to errors the way Python does to exceptions. Here's an example: https://play.golang.org/p/8HZVm8-egIu

There aren't any tracebacks by default, though a `panic()` will show some traceback information. When an error is logged but doesn't panic, there is no traceback. So, for example, an error like "file not found" doesn't give the operator any clues as to what file doesn't exist under what context. If that error says "myFuncName: file %s not found", yields more clues for debugging it. To make adding this context less painful, [errors.Wrap()](https://godoc.org/github.com/pkg/errors#Wrap) (and `Wrapf()`) take a context string and an error and concat them together.

This article is a bit old, but look at the "Annotating Errors" section on https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully; it explains this better than I am :)

Either way, not a blocker for this PR, as returning naked errors (and errors with insufficient context) is all over this codebase and they should probably all be cleaned up.

File: cert/spec.go (hunk 2 above)
Comment (willbuckner) [reply cert/spec.go]: Yeah, I don't have any objective reason other than it's convention. 

```
if err != nil {
    //...
    return
}
``` is such commonly-repeated Go boilerplate that I just tend to notice any time it's different.

**Pull request merged**